src/game/savegame.c

```diff
@@ -493,229 +580,347 @@ void ExtractSaveGameInfo()
-    SkipSG(4);
-
-    ReadSG(&lot->head, sizeof(int16_t));
-    ReadSG(&lot->tail, sizeof(int16_t));
-    ReadSG(&lot->search_number, sizeof(uint16_t));
-    ReadSG(&lot->block_mask, sizeof(uint16_t));
-    ReadSG(&lot->step, sizeof(int16_t));
-    ReadSG(&lot->drop, sizeof(int16_t));
-    ReadSG(&lot->fly, sizeof(int16_t));
-    ReadSG(&lot->zone_count, sizeof(int16_t));
-    ReadSG(&lot->target_box, sizeof(int16_t));
-    ReadSG(&lot->required_box, sizeof(int16_t));
-    ReadSG(&lot->target, sizeof(PHD_VECTOR));
+    SaveGame_SkipSG(4);
+
+    SaveGame_ReadSG(&lot->head, sizeof(int16_t));
+    SaveGame_ReadSG(&lot->tail, sizeof(int16_t));
+    SaveGame_ReadSG(&lot->search_number, sizeof(uint16_t));
+    SaveGame_ReadSG(&lot->block_mask, sizeof(uint16_t));
+    SaveGame_ReadSG(&lot->step, sizeof(int16_t));
+    SaveGame_ReadSG(&lot->drop, sizeof(int16_t));
+    SaveGame_ReadSG(&lot->fly, sizeof(int16_t));
+    SaveGame_ReadSG(&lot->zone_count, sizeof(int16_t));
+    SaveGame_ReadSG(&lot->target_box, sizeof(int16_t));
+    SaveGame_ReadSG(&lot->required_box, sizeof(int16_t));
+    SaveGame_ReadSG(&lot->target, sizeof(PHD_VECTOR));
+}
+
+int16_t SaveGame_LoadSaveBufferFromFile(GAME_INFO *game_info, int32_t slot)
+{
+    assert(game_info);
+
+    char filename[80];
+    sprintf(filename, g_GameFlow.save_game_fmt, slot);
+    LOG_DEBUG("%s", filename);
+
+    MYFILE *fp = File_Open(filename, FILE_OPEN_READ);
+    if (!fp) {
+        return -1;
+    }
+    File_Read(
+        &game_info->savegame_buffer[0], sizeof(char), MAX_SAVEGAME_BUFFER, fp);
+    File_Close(fp);
+
+    SaveGame_ResetSG(game_info);
+    SaveGame_SkipSG(SAVEGAME_TITLE_SIZE);
+    SaveGame_SkipSG(sizeof(int32_t));
+    for (int i = 0; i < g_GameFlow.level_count; i++) {
+        SaveGame_SkipSG(sizeof(uint16_t));
+        SaveGame_SkipSG(sizeof(uint16_t));
+        SaveGame_SkipSG(sizeof(uint16_t));
+        SaveGame_SkipSG(sizeof(uint16_t));
+        SaveGame_SkipSG(sizeof(uint8_t));
+        SaveGame_SkipSG(sizeof(uint8_t));
+        SaveGame_SkipSG(sizeof(uint8_t));
+        SaveGame_SkipSG(sizeof(int8_t));
+        SaveGame_SkipSG(sizeof(int8_t));
+        SaveGame_SkipSG(sizeof(uint16_t));
+    }
+    SaveGame_SkipSG(sizeof(uint32_t));
+    SaveGame_SkipSG(sizeof(uint32_t));
+    SaveGame_SkipSG(sizeof(uint16_t));
+
+    uint16_t level_num;
+    SaveGame_ReadSG(&level_num, sizeof(int16_t));
+
+    return level_num;
+}
+
+bool SaveGame_SaveToFile(GAME_INFO *game_info, int32_t slot)
+{
+    assert(game_info);
+    SaveGame_FillSaveBuffer(game_info);
+
+    char filename[80];
+    sprintf(filename, g_GameFlow.save_game_fmt, slot);
+    LOG_DEBUG("%s", filename);
+
+    MYFILE *fp = File_Open(filename, FILE_OPEN_WRITE);
+    if (!fp) {
+        return false;
+    }
+    File_Write(&game_info->savegame_buffer[0], sizeof(char), m_SGBufPos, fp);
+    File_Close(fp);
+
+    REQUEST_INFO *req = &g_LoadSaveGameRequester;
+    req->item_flags[slot] &= ~RIF_BLOCKED;
+    sprintf(
```

Comment:
What does this do?

Reply:
After my changes it wrongfully updates the saved levels table in the passport with the level filename. Corrected to use level title instead :)